Correct gpio driver

felipebalbi opened the pull request:
Work in progress

Commit:
Correct gpio driver

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

MathisDeroo commented:
Hi @felipebalbi,

IO are controlled by GPIO, PORTx and pcrX registers. GPIO access is already supported. In my driver i have added the PORTx support by adding another field in impl_pin!(P0_0, 0, 0, Gpio0, **Port0**); and did the required modifications.

Looking for guidance on how to do it for pcrX ? For instance P3_18 refers to PORT3 pcr18.
Should I follow the same mechanism of adding another field in impl_pin! like impl_pin!(P3_18, 3, 18, Gpio3, Port3, **Port3::pcr18**); ?
Should I access it with a pointer depending on the pin index ?

felipebalbi commented:
> Hi @felipebalbi,
> 
> IO are controlled by GPIO, PORTx and pcrX registers. GPIO access is already supported. In my driver i have added the PORTx support by adding another field in impl_pin!(P0_0, 0, 0, Gpio0, **Port0**); and did the required modifications.
> 
> Looking for guidance on how to do it for pcrX ? For instance P3_18 refers to PORT3 pcr18. Should I follow the same mechanism of adding another field in impl_pin! like impl_pin!(P3_18, 3, 18, Gpio3, Port3, **Port3::pcr18**); ? Should I access it with a pointer depending on the pin index ?

I think we have the `paste` crate as a dependency on the HAL, you can rely on `paste` to generate those two fields for you from `3` and `18` arguments. Here's an [example ](https://github.com/OpenDevicePartnership/embassy-imxrt/blob/main/src/uart.rs#L1211-L1231) how to use `paste`.

MathisDeroo commented:
> > Hi @felipebalbi,
> > IO are controlled by GPIO, PORTx and pcrX registers. GPIO access is already supported. In my driver i have added the PORTx support by adding another field in impl_pin!(P0_0, 0, 0, Gpio0, **Port0**); and did the required modifications.
> > Looking for guidance on how to do it for pcrX ? For instance P3_18 refers to PORT3 pcr18. Should I follow the same mechanism of adding another field in impl_pin! like impl_pin!(P3_18, 3, 18, Gpio3, Port3, **Port3::pcr18**); ? Should I access it with a pointer depending on the pin index ?
> 
> I think we have the `paste` crate as a dependency on the HAL, you can rely on `paste` to generate those two fields for you from `3` and `18` arguments. Here's an [example ](https://github.com/OpenDevicePartnership/embassy-imxrt/blob/main/src/uart.rs#L1211-L1231) how to use `paste`.

Got it. I assume I still have to use the following impl_pin!(P3_18, 3, 18, Gpio3, Port3, **Port3::pcr18**); but use paste! in the impl_pin to populate the pcrX ?

felipebalbi commented:
> > > Hi @felipebalbi,
> 
> > > IO are controlled by GPIO, PORTx and pcrX registers. GPIO access is already supported. In my driver i have added the PORTx support by adding another field in impl_pin!(P0_0, 0, 0, Gpio0, **Port0**); and did the required modifications.
> 
> > > Looking for guidance on how to do it for pcrX ? For instance P3_18 refers to PORT3 pcr18. Should I follow the same mechanism of adding another field in impl_pin! like impl_pin!(P3_18, 3, 18, Gpio3, Port3, **Port3::pcr18**); ? Should I access it with a pointer depending on the pin index ?
> 
> > 
> 
> > I think we have the `paste` crate as a dependency on the HAL, you can rely on `paste` to generate those two fields for you from `3` and `18` arguments. Here's an [example ](https://github.com/OpenDevicePartnership/embassy-imxrt/blob/main/src/uart.rs#L1211-L1231) how to use `paste`.
> 
> 
> 
> Got it. I assume I still have to use the following impl_pin!(P3_18, 3, 18, Gpio3, Port3, **Port3::pcr18**); but use paste! in the impl_pin to populate the pcrX ?

Not exactly. "Port" and "::pcr" will never change.  You already have 3 and 18 as arguments. Within the macro you use paste to concatenate the full identifier `crate::pac::Port<n>::pcr<pin>` or something like that; can't remember the format off the top of my head.

MathisDeroo commented:
> > > > Hi @felipebalbi,
> > 
> > 
> > > > IO are controlled by GPIO, PORTx and pcrX registers. GPIO access is already supported. In my driver i have added the PORTx support by adding another field in impl_pin!(P0_0, 0, 0, Gpio0, **Port0**); and did the required modifications.
> > 
> > 
> > > > Looking for guidance on how to do it for pcrX ? For instance P3_18 refers to PORT3 pcr18. Should I follow the same mechanism of adding another field in impl_pin! like impl_pin!(P3_18, 3, 18, Gpio3, Port3, **Port3::pcr18**); ? Should I access it with a pointer depending on the pin index ?
> > 
> > 
> > > 
> > 
> > 
> > > I think we have the `paste` crate as a dependency on the HAL, you can rely on `paste` to generate those two fields for you from `3` and `18` arguments. Here's an [example ](https://github.com/OpenDevicePartnership/embassy-imxrt/blob/main/src/uart.rs#L1211-L1231) how to use `paste`.
> > 
> > 
> > Got it. I assume I still have to use the following impl_pin!(P3_18, 3, 18, Gpio3, Port3, **Port3::pcr18**); but use paste! in the impl_pin to populate the pcrX ?
> 
> Not exactly. "Port" and "::pcr" will never change. You already have 3 and 18 as arguments. Within the macro you use paste to concatenate the full identifier `crate::pac::Port<n>::pcr<pin>` or something like that; can't remember the format off the top of my head.

I have added the 'paste' method in the impl_pin as follow, when impl_pin!(P3_18, 3, 18, Gpio3) (removed PortX amd pctX):
 let port_reg = unsafe {&*crate::pac::[<Port $port>]::ptr()};
 port_reg.[<pcr $pin>]().modify(|_, w| w.pe().pe0());

How to do the same in the AnyPin implementation (impl SealedPin for AnyPin) as paste! works at compile time with literal identifiers, that are not exposed to AnyPin ?  In other words PortX and pcrX Are not accessible when AnyPin implements the SealedPin trait.

Commit:
Simplify blinky example

Make it look like every other HAL for consistency. While at that, also
rename the example to match the name used by other HALs.

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

jamesmunns commented on examples/src/bin/blinky.rs:
```diff
@@ -0,0 +1,40 @@
+    defmt::info!("Blink example");
+
+    // Initialize embassy-time global driver backed by OSTIMER0
+    hal::ostimer::time_driver::init(hal::config::Config::default().time_interrupt_priority, 1_000_000);
```
Nit (and not for this PR), but we should probably just move this into `hal::init()` everywhere.

felipebalbi replied:
I have this in another branch

jamesmunns commented on src/lib.rs:
```diff
@@ -14,7 +14,301 @@ pub mod lpuart;
 
-embassy_hal_internal::peripherals!(PORT1, PORT2, PORT3, LPUART2, OSTIMER0, GPIO, PIO2_2, PIO2_3, GPIO3, RTC0, ADC1,);
+#[rustfmt::skip]
+embassy_hal_internal::peripherals!(
```
Just wondering: Is this a comprehensive list now?

felipebalbi replied:
yes, it is. AFAICT, everything is there.

jamesmunns commented on src/gpio.rs:
```diff
@@ -165,6 +354,14 @@ impl<'d> Flex<'d> {
         !self.is_high()
     }
+
+    pub fn is_set_high(&self) -> bool {
```
might be worth adding `#[inline]` to these (and any other public + trivial funcs)

felipebalbi replied:
Done

jamesmunns reviewed (approved):
Overall looks reasonable to me! A couple comments along the way.

We might consider filling in the module/fn level docs a bit, but up to you if that's now or later.

Commit:
Add some documentation to GPIO driver

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

felipebalbi commented:
> Overall looks reasonable to me! A couple comments along the way.
> 
> We might consider filling in the module/fn level docs a bit, but up to you if that's now or later.

Some of this done as well.

Commit:
Enable GPIO clocks during HAL initialization

Provide the user with working GPIO clocks.

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>
